docs/references/http-gateway-protocol-spec.md

```diff
@@ -0,0 +1,471 @@
+
+- The HTTP response status code is taken from the status_code field.
+- The HTTP response headers are taken from the headers field.
+- The HTTP response body is initialized with the value of the body field and further assembled as per the streaming protocol.
+
+Notes:
+
+- Not all HTTP Gateway implementations may be able to pass on all forms of headers. In particular, Service Workers are unable to pass on [forbidden headers](https://fetch.spec.whatwg.org/#forbidden-header-name).
+- HTTP Gateways may add additional headers. In particular, the following headers may be set:
+  - `access-control-allow-origin: \*`
+  - `access-control-allow-methods: GET, POST, HEAD, OPTIONS`
+  - `access-control-allow-headers: DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Cookie`
+  - `access-control-expose-headers: Content-Length,Content-Range`
+  - `x-cache-status: MISS`
+
+## Response Verification
+
+The HTTP Gateway will primarily be used to load static assets needed to run frontend canister code, so both low latency and security are essential for providing a good experience to end users. [Query calls](https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-query) are more performant but less secure than [Update calls](https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-call).
+
+Response verification fills the security gap left by query calls. It is a versioned subprotocol that allows for an HTTP Gateway to verify a certified response received as a result of performing a query call to the Internet Computer. Two versions are currently supported, the older version 1 only supports verification of the response body, whereas the newer version 2 supports verifying the full response, including headers, status code, and optionally the full request.
+
+Response verification comes in two flavors, the current version is covered in this section and the legacy version is covered in [another section](#legacy-response-verification). The legacy version of response verification only includes a mapping of the request URL to the response body so it is quite limiting in what it can verify. The current version builds on the legacy version by optionally including the following extra parameters in the certification process:
+
+- Request URL query params
+- Request method
+- Request headers
+- Response status code
+- Response headers
+
+### Response Verification Outline
+
+1. Case-insensitive search for the `IC-Certificate` response header.
+   - If no such header is found, verification fails.
+   - If the header value is not structured as per [the certificate header](#the-certificate-header), verification fails.
+1. Parse the `certificate`, `tree`, and `expr_path` fields from the `IC-Certificate` header value as per [the certificate header](#the-certificate-header).
+1. Case-insensitive search for the `IC-CertificationExpression` header.
```

Comment:
Shouldn't this depend on the `version` in the step after the next?

Reply:
You're right, I'll reorder these, thanks